torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py

```diff
@@ -0,0 +1,330 @@
+
+    Args: same as ``torch.Tensor.type_as``.
+
+    Return:
+        new_local_shards (List[Shard]): Local shards for the new sharded tensor.
+        st_meta (ShardedTensorMetadata): Metadata of the new sharded tensor.
+    """
+    st = args[0]
+    tensor = args[1]
+    if isinstance(tensor, ShardedTensor):
+        tensor = tensor.local_tensor()
+    new_local_shards = []
+    for shard in st.local_shards():
+        new_local_shards.append(Shard(shard.tensor.type_as(tensor), shard.metadata))
+    st_meta = copy.deepcopy(st._metadata)
+    st_meta.tensor_properties.dtype = tensor.dtype
+    return new_local_shards, st_meta
+
+
+_register_sharded_op_on_local_shards(
+    torch.Tensor.type_as,
+    early_stop_func=same_type,
+    extra_check=sharded_type_as_check,
+    customized_func=sharded_type_as,
+)
+
+
+def transpose_same_dim(*args, **kwargs):
+    """
+    When the dim0 and dim1 of transpose are the same, return the original ShardedTensor.
+
+    Args: same as ``torch.Tensor.transpose``.
+
+    Return (bool): Whether to return early or not.
+    """
+    return args[1] == args[2]
+
+
+def sharded_transpose_check(*args, **kwargs):
+    """
+    Perform extra checks for the sharded_transpose op such as the input needs to
+    be at least 2 and the sharding spec needs to be a ChunkShardingSpec.
+
+    Args: same as ``torch.Tensor.type_as``.
+
+    Return: None
+    """
+    if len(args) < 3:
+        raise ValueError("Needs at least two dimensions for transpose op!")
+    _chunk_sharding_spec_check(args[0].sharding_spec(), torch.Tensor.transpose)
+
+
+def sharded_transpose(args, kwargs, pg):
+    """
+    Handles ``__torch_function__`` dispatch for the sharded_transpose op.
+
+    Returns a new sharded tensor with the given dimensions transposed.
+    During the transpose, we keep the original shading dim, if the sharding
+    dim is not neither dim0 nor dim1. Otherwise, we will swap the sharding
+    dim with the other input of transpose.
+
+    Args: (same as ``torch.Tensor.transpose``.)
+        dim0 (Int): the first dimension to be transposed.
+        dim1 (Int): the second dimension to be transposed.
+
+    Returns:
+        new_local_shards (List[Shard]): Local shards for the new sharded tensor.
+        st_meta (ShardedTensorMetadata): Metadata of the new sharded tensor.
+    """
+
+    def _swap_meta_data(data, idx0, idx1):
+        """
+        Swap the item at idx0 and idx1 in the data list.
+        """
+        data[idx0], data[idx1] = data[idx1], data[idx0]
+
+    st = args[0]
+    dim0 = args[1]
+    dim1 = args[2]
+
+    new_local_shards = []
+    for shard in st.local_shards():
+        shard_meta_data = copy.deepcopy(shard.metadata)
+        _swap_meta_data(shard_meta_data.shard_offsets, dim0, dim1)
+        _swap_meta_data(shard_meta_data.shard_sizes, dim0, dim1)
+        new_local_shards.append(
+            Shard(shard.tensor.transpose(dim0, dim1).contiguous(), shard_meta_data)
```

Comment:
should we insert the `contiguous()` call implicitly? or we rely on user to call `contiguous()` after the `transpose`? I feel like we shouldn't do it implicitly here unless we have a reason to

Reply:
I agree but Shard here will throw exception if we don't call `contiguous` here..